### PR TITLE
Indexes and constraints are now baked with the snapshot

### PR DESCRIPTION
**Pull request by HavokInspiration**
When baking a snapshot, indexes and constraints where not baked and thus, lost.
This PR aims to correct that.

The only thing I could not do was to support composite foreign keys as Cake doesn't seem to support them. Am I correct ? From what I gathered, when a table has a multi-columns foreign key defined and is reflected, the last column within the list is the one that persists.
Should I create an issue in the main repo with what I used to test and conclude this ?

More details on #80

### CODE REVIEW

**Comment by HavokInspiration** [comment]
I'll try to fix the build later.

**Comment by markstory** [comment]
Multi-colum constraints should be working in cakephp/cakephp. If the aren't we should fix that as it will be a problem for fixture creation/use.

**Comment by markstory** [comment]
Looks good :+1:

**Comment by lorenzo** [comment]
@HavokInspiration what is the problem with composite primary keys?

**Comment by HavokInspiration** [comment]
@lorenzo Composite primary keys works ok. Composite foreign keys don't. Only the last referenced column is returned when the table is reflected.
When I get the time, I'll post a full issue on the cakephp repository with everything that led to me to conclude that there might be a problem.

**Comment by HavokInspiration** [comment]
I just pushed another commit to try to resolve current failure. Locally, everything is passing.
My problem was that, with MySQL, foreign keys were not referenced as indexes when reflecting the table, which is what was causing differences in the output expectation.

So before printing a ``addIndex`` statement, I compare the columns list to the primary key list : if they are the same, no need to have a ``addIndex`` statement. Same thing goes for constraints.

I don't really know if that makes much sense... It was however, the only way I found to avoid having differences in the output for different drivers.
I haven't had the chance to migrate the resulting file to see if I get the same file after taking a snapshot of the first migration. Maybe it's worth doing it before merging.

(Out of curiosity, do you know a way to avoid the timestamp comparison test to sometimes fail, and sometimes pass ?)

File: src/Template/Bake/config/snapshot.ctp [hunk 1]
```diff
@@ -51,6 +53,42 @@ class <%= $name %> extends AbstractMigration
                 echo $this->Migration->stringifyList($columnOptions, ['indent' => 4]);
             %>])
         <%- endforeach; %>
+        <%- foreach ($this->Migration->constraints($table) as $constraint):
+            $constraintColumns = $constraint['columns'];
+            sort($constraintColumns);
+            if ($constraint['type'] !== 'unique'):
+                $foreignKeys[] = $constraint['columns'][0]; %>
```
Comment (HavokInspiration) [inline src/Template/Bake/config/snapshot.ctp]: This (and above) needs some clean up

File: src/Template/Bake/config/snapshot.ctp [hunk 2]
```diff
@@ -51,6 +53,42 @@ class <%= $name %> extends AbstractMigration
...
+                    echo $this->Migration->stringifyList($constraint['columns'], ['indent' => 5]);
+                %>],
+                ['unique' => true]
+            )
+            <%- endif; %>
+            <%- endforeach; %>
+        <%- foreach($this->Migration->indexes($table) as $index):
+            sort($foreignKeys);
+            $indexColumns = $index['columns'];
+            sort($indexColumns);
```
Comment (HavokInspiration) [inline src/Template/Bake/config/snapshot.ctp]: Clean up needed here too.

**Comment by dereuromark** [comment]
Those are 1second-leap-issues occuring when the seconds move on during the execution. Nothing bad, just annoying.
So all good :+1:

**Comment by lorenzo** [comment]
Thanks @HavokInspiration !

**Pull request merged**